Fix #544: Correctly send parameters passed to NormandyApi.

Osmose opened the pull request:
Updates NormandyApi.jsm to send request parameters correctly as query
string arguments for GET requests and as JSON-encoded data in the
request body for POST requests.

gijsk commented on recipe-client-addon/lib/NormandyApi.jsm:
```diff
@@ -10,35 +9,32 @@ Cu.import("resource://gre/modules/Services.jsm");
-        for (const key in data) {
-          paramObj.append(key, data[key]);
+    let body = undefined;
+    if (data !== undefined) {
```
Is there any reason not to just use:

```js
if (data) {
  ...
}
```

?

gijsk commented on recipe-client-addon/test/browser/test_server.sjs:
```diff
@@ -1,8 +1,30 @@
+  const bytes = [];
+  let available;
+  while ((available = inputStream.available()) > 0) {
+    Array.prototype.push.apply(bytes, inputStream.readByteArray(available));
```
I think you can just use:

```js
bytes = bytes.concat(inputStream.readByteArray(available));
```
which is a little easier?

gijsk commented on recipe-client-addon/test/browser/test_server.sjs:
```diff
@@ -1,8 +1,30 @@
+  const params = request.queryString.split("&");
+  for (const param of params) {
+    const [key, value] = param.split("=");
+    data.queryString[key] = value;
```
Should this unescape? I guess it's just a test, so maybe it doesn't matter...

Osmose replied:
Yeah, I don't want to bother writing a robust fake server.

gijsk commented on recipe-client-addon/test/browser/browser_NormandyApi.js:
```diff
@@ -15,7 +15,23 @@ add_task(function* () {
-  Assert.deepEqual(data, {test: "data"}, "NormandyApi returned incorrect server data.");
+  let response = yield NormandyApi.get("browser/test_server.sjs");
+  let data = yield response.json();
+  Assert.deepEqual(data, {queryString: {}, body: {}}, "NormandyApi returned incorrect server data.");
```
Is this going to conflict with the other patch here?

Osmose replied:
I suspect it will. Can you use `sjs` files in xpcshell tests? All the docs I found on them were mochitest docs.

gijsk replied:
Yes, I think the new test in #510 uses one.

mythmon replied:
#510 doesn't use any `.sjs` files, but you should  be able to, according to [the httpd docs](https://developer.mozilla.org/en-US/docs/Mozilla/httpd.js/HTTP_server_for_unit_tests#SJS:_server-side_scripts). You might have to add `server.registerContentType("sjs", "sjs");`, but I wouldn't be surprised if that were the default.

gijsk commented on recipe-client-addon/lib/NormandyApi.jsm:
```diff
@@ -10,35 +9,32 @@ Cu.import("resource://gre/modules/Services.jsm");
+    if (data !== undefined) {
+      if (method === "get") {
+        for (const key of Object.keys(data)) {
+          url.searchParams.set(key, data[key]);
```
Sadfaces that there's no batch-set API here...

Osmose commented:
Addressed feedback and moved the tests to the xpcshell tests. I had to fix some issues with the existing xpcshell tests since they appeared to not be running properly, for some reason.
